pkg/services/ngalert/api/api_ruler.go

```diff
@@ -87,68 +90,65 @@ func (srv RulerSrv) RouteDeleteAlertRules(c *models.ReqContext, namespaceTitle s
-		// being the rules that fulfill the predicate the other partation being
-		// the ruleIDs not fulfilling it.
-		partition := func(alerts []*ngmodels.AlertRule, predicate func(rule *ngmodels.AlertRule) bool) ([]*ngmodels.AlertRule, []string) {
-			positive, negative := make([]*ngmodels.AlertRule, 0, len(alerts)), make([]string, 0, len(alerts))
-			for _, rule := range alerts {
-				if predicate(rule) {
-					positive = append(positive, rule)
-					continue
+		var groupsToDelete = make(map[ngmodels.AlertRuleGroupKey][]*ngmodels.AlertRule)
+		for _, rule := range q.Result {
+			key := rule.GetGroupKey()
+			groupsToDelete[key] = append(groupsToDelete[key], rule)
+		}
+		rulesToDelete := make([]string, 0, len(q.Result))
+		for groupKey, rules := range groupsToDelete {
+			if !authorizeAccessToRuleGroup(rules, hasAccess) {
+				unauthz = true
+				break
+			}
+			canDelete := true
+			uid := make([]string, 0, len(rules))
+			keys := make([]ngmodels.AlertRuleKey, 0, len(rules))
+			for _, rule := range rules {
+				provenance, exists := provenances[rule.UID]
+				// if at least one rule is provisioned do not allow to delete via regular api
+				if exists && provenance != ngmodels.ProvenanceNone {
+					provisioned = true
+					canDelete = false
+					break
 				}
-				negative = append(negative, rule.UID)
+				uid = append(uid, rule.UID)
+				keys = append(keys, rule.GetKey())
+			}
+			if canDelete {
+				rulesToDelete = append(rulesToDelete, uid...)
+				deletedGroups[groupKey] = keys
 			}
-			return positive, negative
-		}
-
-		canDelete, cannotDelete = partition(q.Result, func(rule *ngmodels.AlertRule) bool {
-			return authorizeDatasourceAccessForRule(rule, hasAccess)
-		})
-		if len(canDelete) == 0 {
-			return fmt.Errorf("%w to delete rules because user is not authorized to access data sources used by the rules", ErrAuthorization)
-		}
-		if len(cannotDelete) > 0 {
-			logger.Info("user cannot delete one or many alert rules because it does not have access to data sources. Those rules will be skipped", "expected", len(q.Result), "authorized", len(canDelete), "unauthorized", cannotDelete)
 		}
-
-		canDelete, cannotDelete = partition(canDelete, func(rule *ngmodels.AlertRule) bool {
-			provenance, exists := provenances[rule.UID]
-			return (exists && provenance == ngmodels.ProvenanceNone) || !exists
-		})
-
-		if len(canDelete) == 0 {
-			return fmt.Errorf("all rules have been provisioned and cannot be deleted through this api")
+		if len(rulesToDelete) > 0 {
+			return srv.store.DeleteAlertRulesByUID(ctx, c.SignedInUser.OrgID, rulesToDelete...)
 		}
-
-		if len(cannotDelete) > 0 {
-			logger.Info("user cannot delete one or many alert rules because it does have a provenance set. Those rules will be skipped", "expected", len(q.Result), "provenance_none", len(canDelete), "provenance_set", cannotDelete)
+		// if none rules were deleted return an error.
+		if provisioned {
+			return errProvisionedResource
 		}
```

Comment:
Partial errors are always difficult to represent in APIs.. do we have a way of telling the user which of their requests succeeded and which failed?

If we can't do this, could we at a minimum keep the log statement that happens when an object gets skipped but the request is not accepted?

Reply:
It is not that partial, actually. 
1. when a single group is being deleted. The permissions to perform the delete action in the folder are checked before this code, on the router level. Therefore, this handler needs to check whether the group is accessible to the user. If it is not, then provisioning status will not be checked at all and therefore the response will be 401. If the user does have access to the group but it is provisioned - the response is 400 due to the status of alerts.
2. "all" rules in the folder are deleted. I used quotes because "all" depends on the user permissions. For example, folder contains 4 groups:
- "group1" that references data sources to which the user has permissions, 
- "group2". It uses data sources to which the user does not have permission.
- "group3". Provisioned rules which user is authorized to see.
- "group4". Provisioned rules which user is not authorized to see.
Now when the user requests the list of rules in that folder, the API returns just two groups "group1" and "group3". 
2.1. Now the user tries to delete rules in that folder. The API will read 4 groups and will remove the "group1", and the result is 202
2.2. Now the user sees only the provisioned group "group3". It tries to delete the folder again. The API will read 3 groups and will exclude 2 groups that user does not have access to (and not visibile to it) and will fail with the code 400.
2.3. Let's say user deleted "group3" via provisioning API. Although the GET API will not return the rules in that folder, the user remembers that the folder exsits and tries to delete all rules in the folder again. The handler will read 2 groups, and will return 401. The provisioning status will not be checked at all.